### PR TITLE
Closes #37 | Create dataset loader for MEmoLon

### LINKED ISSUE #37: Create dataset loader for MEmoLon
Dataloader name: `memolon/memolon.py`
DataCatalogue: http://seacrowd.github.io/seacrowd-catalogue/card.html?memolon

|     Dataset| memolon |
|-------------|---|
| Description | MEmoLon is an emotion lexicons for 91 languages, each one covers eight emotional variables and comprises over 100k word entries. There are several versions of the lexicons, the difference being the choice of the expansion model |
| Subsets | - |
| Languages | ind, zlm, tgl, vie, sun, tha, mya, ceb, jav |
| Tasks | Emotion Classification |
| License | MIT (mit) |
| Homepage | https://zenodo.org/record/3756607/files/MTL_grouped.zip?download=1 |
| HF URL | - |
| Paper URL | https://aclanthology.org/2020.acl-main.112.pdf |

### PR DESCRIPTION
**Pull request by danjohnvelasco**
Closes #37 

### Checkbox
- [X] Confirm that this PR is linked to the dataset issue.
- [X] Create the dataloader script `seacrowd/sea_datasets/my_dataset/my_dataset.py` (please use only lowercase and underscore for dataset naming).
- [X] Provide values for the `_CITATION`, `_DATASETNAME`, `_DESCRIPTION`, `_HOMEPAGE`, `_LICENSE`, `_URLs`, `_SUPPORTED_TASKS`, `_SOURCE_VERSION`, and `_SEACROWD_VERSION` variables.
- [X] Implement `_info()`, `_split_generators()` and `_generate_examples()` in dataloader script.
- [ ] Make sure that the `BUILDER_CONFIGS` class attribute is a list with at least one `SEACrowdConfig` for the source schema and one for a seacrowd schema.
- [X] Confirm dataloader script works with `datasets.load_dataset` function.
- [ ] Confirm that your dataloader script passes the test suite run with `python -m tests.test_seacrowd seacrowd/sea_datasets/<my_dataset>/<my_dataset>.py`.
- [ ] If my dataset is local, I have provided an output of the unit-tests in the PR (please copy paste). This is OPTIONAL for public datasets, as we can test these without access to the data files.

> Make sure that the `BUILDER_CONFIGS` class attribute is a list with at least one `SEACrowdConfig` for the source schema and one for a seacrowd schema.

For this dataset, only source schema will be implemented (see #37)

> Confirm that your dataloader script passes the test suite run with `python -m tests.test_seacrowd seacrowd/sea_datasets/<my_dataset>/<my_dataset>.py`.

Fails not having `<dataset_name>_source` in config. This dataset have multiple languages/subsets and config is named as `<dataset_name>_<lang>_source`.

### CODE REVIEW

**Comment by sabilmakbar** [comment]
> Fails not having <dataset_name>_source in config. This dataset have multiple languages/subsets and config is named as <dataset_name>_<lang>_source.

Hi @danjohnvelasco, in that case, can you try to do this?
```bash
python -m tests.test_seacrowd seacrowd/sea_datasets/<your_dataset>/<your_dataset>.py --subset_id <your_dataset>_<lang>
```

It will check your implementation on `<your_dataset>_<lang>_source` and `<your_dataset>_<lang>_seacrowd_<schema_suffix>`, w/ `<schema_suffix>` is/are chosen based on your `_SUPPORTED_TASKS` variable using `TASK_TO_SCHEMA` var in `utils/constants.py`

**Comment by danjohnvelasco** [comment]
Hi @sabilmakbar, thanks for the suggestion :)

I ran the source only test using the ff. command for each language:
```
python -m tests.test_seacrowd_source_only seacrowd/sea_datasets/memolon/memolon.py --subset_id memolon_<lang>
```

The tests are passing for all supported languages (omitted some outputs and warnings for brevity)
```
INFO:__main__:Checking load_dataset with config name memolon_en_source

Generating train split: 2000004 examples [03:18, 10080.10 examples/s]

INFO:__main__:Dataset sample [source]
{'word': 'aardvark', 'valence': 5.300000190734863, 'arousal': 3.5799999237060547, 'dominance': 5.050000190734863, 'joy': 2.559999942779541, 'anger': 1.3200000524520874, 'sadness': 1.2999999523162842, 'fear': 1.3700000047683716, 'disgust': 1.2899999618530273}
.
----------------------------------------------------------------------
Ran 1 test in 200.096s

OK

INFO:__main__:Checking load_dataset with config name memolon_ceb_source

Generating train split: 2006001 examples [02:46, 12022.21 examples/s]

INFO:__main__:Dataset sample [source]
{'word': 'aardvark', 'valence': 4.949999809265137, 'arousal': 4.179999828338623, 'dominance': 5.090000152587891, 'joy': 1.9199999570846558, 'anger': 1.6200000047683716, 'sadness': 1.5700000524520874, 'fear': 1.649999976158142, 'disgust': 1.6100000143051147}
.
----------------------------------------------------------------------
Ran 1 test in 168.193s

OK

INFO:__main__:Checking load_dataset with config name memolon_tl_source

INFO:__main__:Dataset sample [source]
{'word': 'aardvark', 'valence': 4.409999847412109, 'arousal': 3.950000047683716, 'dominance': 4.659999847412109, 'joy': 1.8799999952316284, 'anger': 1.559999942779541, 'sadness': 1.5099999904632568, 'fear': 1.590000033378601, 'disgust': 1.559999942779541}
.
----------------------------------------------------------------------
Ran 1 test in 0.582s

OK


INFO:__main__:Checking load_dataset with config name memolon_id_source

Generating train split: 2002221 examples [02:44, 12199.99 examples/s]

INFO:__main__:Dataset sample [source]
{'word': 'tapir', 'valence': 5.269999980926514, 'arousal': 3.75, 'dominance': 4.940000057220459, 'joy': 2.4800000190734863, 'anger': 1.340000033378601, 'sadness': 1.3300000429153442, 'fear': 1.3799999952316284, 'disgust': 1.309999942779541}
.
----------------------------------------------------------------------
Ran 1 test in 165.271s

OK

INFO:__main__:Checking load_dataset with config name memolon_su_source

Generating train split: 327533 examples [00:26, 12575.44 examples/s]

INFO:__main__:Dataset sample [source]
{'word': 'aardvark', 'valence': 4.800000190734863, 'arousal': 4.28000020980835, 'dominance': 4.989999771118164, 'joy': 1.8600000143051147, 'anger': 1.6699999570846558, 'sadness': 1.6100000143051147, 'fear': 1.7000000476837158, 'disgust': 1.659999966621399}
.
----------------------------------------------------------------------
Ran 1 test in 28.061s

OK

INFO:__main__:Checking load_dataset with config name memolon_jv_source

Generating train split: 518634 examples [00:42, 12329.49 examples/s]

INFO:__main__:Dataset sample [source]
{'word': 'pakaryan', 'valence': 5.550000190734863, 'arousal': 3.880000114440918, 'dominance': 5.550000190734863, 'joy': 2.059999942779541, 'anger': 1.409999966621399, 'sadness': 1.3700000047683716, 'fear': 1.4299999475479126, 'disgust': 1.399999976158142}
.
----------------------------------------------------------------------
Ran 1 test in 44.061s

OK

INFO:__main__:Checking load_dataset with config name memolon_ms_source

Generating train split: 1213397 examples [01:37, 12387.04 examples/s]

INFO:__main__:Dataset sample [source]
{'word': 'aardvark', 'valence': 5.099999904632568, 'arousal': 3.7799999713897705, 'dominance': 5.03000020980835, 'joy': 2.0299999713897705, 'anger': 1.4500000476837158, 'sadness': 1.409999966621399, 'fear': 1.4700000286102295, 'disgust': 1.4600000381469727}
.
----------------------------------------------------------------------
Ran 1 test in 99.071s

OK

INFO:__main__:Checking load_dataset with config name memolon_vi_source

Generating train split: 2008605 examples [02:43, 12269.47 examples/s]

INFO:__main__:Dataset sample [source]
{'word': 'trả lời', 'valence': 5.460000038146973, 'arousal': 3.890000104904175, 'dominance': 5.489999771118164, 'joy': 2.4800000190734863, 'anger': 1.4299999475479126, 'sadness': 1.399999976158142, 'fear': 1.440000057220459, 'disgust': 1.440000057220459}
.
----------------------------------------------------------------------
Ran 1 test in 165.831s

OK

INFO:__main__:Checking load_dataset with config name memolon_th_source

Generating train split: 2006540 examples [02:44, 12161.23 examples/s]

INFO:__main__:Dataset sample [source]
{'word': 'มด', 'valence': 4.5, 'arousal': 3.5199999809265137, 'dominance': 4.590000152587891, 'joy': 1.8799999952316284, 'anger': 1.5199999809265137, 'sadness': 1.4600000381469727, 'fear': 1.5399999618530273, 'disgust': 1.5199999809265137}
.
----------------------------------------------------------------------
Ran 1 test in 167.123s

OK

INFO:__main__:Checking load_dataset with config name memolon_my_source

Generating train split: 339628 examples [00:27, 12428.57 examples/s]

INFO:__main__:Dataset sample [source]
{'word': 'ကဗျာ', 'valence': 4.760000228881836, 'arousal': 4.340000152587891, 'dominance': 4.880000114440918, 'joy': 1.8700000047683716, 'anger': 1.649999976158142, 'sadness': 1.5700000524520874, 'fear': 1.6699999570846558, 'disgust': 1.6399999856948853}
.
----------------------------------------------------------------------
Ran 1 test in 28.300s

OK

INFO:__main__:Checking load_dataset with config name memolon_zh_source

Generating train split: 2001799 examples [02:42, 12289.05 examples/s]

INFO:__main__:Dataset sample [source]
{'word': '土豚', 'valence': 4.860000133514404, 'arousal': 3.9100000858306885, 'dominance': 4.800000190734863, 'joy': 2.5799999237060547, 'anger': 1.340000033378601, 'sadness': 1.350000023841858, 'fear': 1.3899999856948853, 'disgust': 1.309999942779541}
.
----------------------------------------------------------------------
Ran 1 test in 165.019s

OK
```

File: seacrowd/sea_datasets/memolon/memolon.py
Comment (jamesjaya) [inline seacrowd/sea_datasets/memolon/memolon.py:82]: please add Dataset suffix here
```suggestion
class MemolonDataset(datasets.GeneratorBasedBuilder):
```

File: seacrowd/sea_datasets/memolon/memolon.py
Comment (jamesjaya) [inline seacrowd/sea_datasets/memolon/memolon.py:66]: - let's put this inside the class?
- typing
```suggestion
def seacrowd_config_constructor(lang: str, schema: str, version: str) -> SEACrowdConfig:
```

File: seacrowd/sea_datasets/memolon/memolon.py
Comment (jamesjaya) [inline seacrowd/sea_datasets/memolon/memolon.py:61]: let's remove en and zh since they are not included in the issue?

File: seacrowd/sea_datasets/memolon/memolon.py
Comment (jamesjaya) [inline seacrowd/sea_datasets/memolon/memolon.py:87]: this must refer to one of the config names, in your case, it is `f"{_DATASETNAME}_{lang}_source"`, maybe just pick one of the language to be here

File: seacrowd/sea_datasets/memolon/memolon.py
Comment (jamesjaya) [inline seacrowd/sea_datasets/memolon/memolon.py:85]:
```suggestion
    BUILDER_CONFIGS = [SEACrowdConfig(name=f"{_DATASETNAME}_{lang}_source", version=datasets.Version(_SOURCE_VERSION), description=f"MEmoLon source schema for {lang} language", schema="source", subset_id="memolon") for lang in _LANGUAGE_MAP]
```

**Review by jamesjaya — CHANGES_REQUESTED** [review]

File: seacrowd/sea_datasets/memolon/memolon.py
Comment (danjohnvelasco) [reply seacrowd/sea_datasets/memolon/memolon.py:82]: Hi @jamesjaya thanks for reviewing my PR. 

Regarding the dataset suffix, if I understand correctly, the instruction is to name the class similar to the python script name but in CamelCase (see below for reference).

https://github.com/SEACrowd/seacrowd-datahub/blob/16116789ff8e0bb5a3c87387586fcae1393af648/templates/template.py#L104-L105

File: seacrowd/sea_datasets/memolon/memolon.py
Comment (danjohnvelasco) [reply seacrowd/sea_datasets/memolon/memolon.py:66]: From what I observed, existing scripts defined `seacrowd_config_constructor` outside the class. I'm wondering if we have decided on a convention on whether to define it inside or outside the class? In my opinion, it would be better if there's such a convention. Please do let me know. 

Citing some scripts below (non-exhaustive list) that I found to have defined `seacrowd_config_constructor` outside the class:
https://github.com/SEACrowd/seacrowd-datahub/blob/16116789ff8e0bb5a3c87387586fcae1393af648/seacrowd/sea_datasets/nllb_seed/nllb_seed.py#L98

https://github.com/SEACrowd/seacrowd-datahub/blob/16116789ff8e0bb5a3c87387586fcae1393af648/seacrowd/sea_datasets/nusax_mt/nusax_mt.py#L57

**Comment by SamuelCahyawijaya** [comment]
@danjohnvelasco , any update for this PR?

File: seacrowd/sea_datasets/memolon/memolon.py
Comment (danjohnvelasco) [reply seacrowd/sea_datasets/memolon/memolon.py]: In my opinion, maybe it's better for multilingual datasets to not have a `DEFAULT_CONFIG_NAME` so I just set it to None.

File: seacrowd/sea_datasets/memolon/memolon.py
Comment (danjohnvelasco) [reply seacrowd/sea_datasets/memolon/memolon.py]: Since there's no response regarding conventions for `seacrowd_config_constructor` or similar, I'll keep it outside the class similar to the approved dataloaders I mentioned above.

For the typing, thanks for the suggestion. I already applied it and pushed the changes.

**Comment by danjohnvelasco** [comment]
> @danjohnvelasco , any update for this PR?

Hi @SamuelCahyawijaya, I accepted some of the requested changes. For those change requests that aren't accepted, I gave my reasoning and/or suggestions. Everything is implemented now and the changes are pushed, just waiting for approval. Thanks!

File: seacrowd/sea_datasets/memolon/memolon.py
Comment (jamesjaya) [reply seacrowd/sea_datasets/memolon/memolon.py:84]: Well, as a reviewer, I have a conflicting instruction to make sure that the class name has `Dataset` suffix. 

Quoting this document https://docs.google.com/document/d/1uVtus5MXeM2lf4MmlSGhSXIHZXVLCqoaK6nh6sMyC3w/edit
```
CamelCase for dataloader class name (and “Dataset” is contained in the suffix of the class name).
```

I'll defer to @SamuelCahyawijaya

**Comment by jamesjaya** [comment]
Hey @SamuelCahyawijaya, please check some of the discussion points. This looks ok to me, I will approve after @SamuelCahyawijaya gives his approval.

**Review by SamuelCahyawijaya — APPROVED** [review]
Thank you @danjohnvelasco for the contribution! Tested & LGTM! cc @jamesjaya

**Pull request merged**
